sdk/src/account_context.rs

```diff
@@ -0,0 +1,258 @@
+            account_is_writable: Vec::with_capacity(instruction_accounts.len()),
+        };
+        for (index, is_signer, is_writable) in instruction_accounts.iter() {
+            result.account_indices.push(*index);
+            result.account_is_signer.push(*is_signer);
+            result.account_is_writable.push(*is_writable);
+        }
+        self.invoke_stack.push(result);
+        Ok(())
+    }
+
+    pub fn invocation_stack_pop(&mut self) -> Result<(), InstructionError> {
+        if self.invoke_stack.is_empty() {
+            return Err(InstructionError::CallDepth);
+        }
+        self.invoke_stack.pop();
+        Ok(())
+    }
+
+    pub fn set_return_data(&mut self, data: Vec<u8>) -> Result<(), InstructionError> {
+        let pubkey = *self.borrow_program_account()?.key();
+        self.return_data = (pubkey, data);
+        Ok(())
+    }
+
+    pub fn get_return_data(&self) -> (&Pubkey, &[u8]) {
+        (&self.return_data.0, &self.return_data.1)
+    }
+}
+
+/// Loaded instruction shared between runtime and programs.
+///
+/// This context is valid for the entire duration of a (possibly cross program) instruction being processed.
+pub struct InstructionContext {
+    number_of_program_accounts: usize,
+    account_indices: Vec<usize>,
+    account_is_signer: Vec<bool>,
+    account_is_writable: Vec<bool>,
+    instruction_data: Vec<u8>,
+}
+
+impl InstructionContext {
+    pub fn number_of_program_accounts(&self) -> usize {
+        self.number_of_program_accounts
+    }
+
+    pub fn number_of_accounts(&self) -> usize {
+        self.account_indices.len()
+    }
+
+    pub fn instruction_data(&self) -> &[u8] {
+        &self.instruction_data
+    }
+
+    pub fn borrow_account<'a, 'b: 'a>(
+        &'a self,
+        transaction_context: &'b TransactionContext,
+        index_in_instruction: usize,
+    ) -> Result<BorrowedAccount<'a>, InstructionError> {
+        if index_in_instruction >= self.account_indices.len() {
+            return Err(InstructionError::NotEnoughAccountKeys);
+        }
+        let index_in_transaction = self.account_indices[index_in_instruction];
+        if index_in_transaction >= transaction_context.accounts.len() {
+            return Err(InstructionError::MissingAccount);
+        }
+        let account = transaction_context.accounts[index_in_transaction].borrow_mut();
+        Ok(BorrowedAccount {
+            transaction_context,
+            instruction_context: self,
+            index_in_transaction,
+            index_in_instruction,
+            account,
+        })
+    }
+}
+
+/// Shared account borrowed from the TransactionContext and an InstructionContext.
+pub struct BorrowedAccount<'a> {
+    transaction_context: &'a TransactionContext,
+    instruction_context: &'a InstructionContext,
+    index_in_transaction: usize,
+    index_in_instruction: usize,
+    account: RefMut<'a, AccountSharedData>,
+}
+
+impl<'a> BorrowedAccount<'a> {
+    pub fn key(&self) -> &Pubkey {
```

Comment:
If this is called a lot might want to do the work once and stick the pubkey into `BorrowAccount`